core/src/main/java/io/grpc/ServiceDescriptor.java

```diff
@@ -108,4 +132,89 @@ static void validateMethodNames(String serviceName, Collection<MethodDescriptor<
           "duplicate name %s", method.getFullMethodName());
     }
   }
+
+  /**
+   * Creates a new builder for a {@link ServiceDescriptor}.
+   *
+   * @since 1.1.0
+   */
+  @ExperimentalApi("https://github.com/grpc/grpc-java/issues/2666")
+  public static Builder newBuilder() {
```

Comment:
I don't think proto is a good comparison, because it could not do the pattern I'm suggesting. And I think with proto people expect to build() to fail the first couple of times they run their program. That is needless in this case. If there are a lot of invariants I'd expect to do something different.

Looking through Guava, I don't see a _single_ Builder that can be in an incomplete state. Their API is designed where you simply can't build it wrong. The main non-trivial example is [GraphBuilder](http://google.github.io/guava/releases/snapshot/api/docs/com/google/common/graph/GraphBuilder.html)/NetworkBuilder/ValueGraphBuilder with directed()/undirected(). There are some older+weaker examples like [ThreadFactoryBuilder](http://google.github.io/guava/releases/snapshot/api/docs/com/google/common/util/concurrent/ThreadFactoryBuilder.html) which falls back to defaults and [MinMaxPriorityQueue.Builder](http://google.github.io/guava/releases/snapshot/api/docs/com/google/common/collect/MinMaxPriorityQueue.Builder.html) which you provide an Iterable to the `create()` (which is basically equivalent to `build()`).